File: app/src/main/java/com/yao/zhihudaily/ui/hot/HotMainFragment.java
package com.yao.zhihudaily.ui.hot;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.orhanobut.logger.Logger;
import com.yao.zhihudaily.R;
import com.yao.zhihudaily.model.Hot;
import com.yao.zhihudaily.model.HotJson;
import com.yao.zhihudaily.net.ZhihuHttp;
import com.yao.zhihudaily.tool.SimpleDividerDecoration;
import com.yao.zhihudaily.ui.BaseFragment;

import java.util.ArrayList;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observer;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;

/**
 * @author Yao
 * @date 2016/7/22
 */
public class HotMainFragment extends BaseFragment {

    private static final String TAG = "HotMainFragment";

    @BindView(R.id.rv_hots)
    RecyclerView mRvHots;

    private HotAdapter hotAdapter;

    private Disposable mDisposable;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_hot, container, false);
        ButterKnife.bind(this, view);

        LinearLayoutManager linearLayoutManager;
        mRvHots.setLayoutManager(new LinearLayoutManager(getFragmentActivity()));
        //mRvHots.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL_LIST));
        mRvHots.addItemDecoration(new SimpleDividerDecoration(getFragmentActivity()));
        mRvHots.setAdapter(hotAdapter = new HotAdapter(this));

        getHot();


        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mDisposable != null) {
            mDisposable.dispose();
        }
    }

    private void getHot() {
        ZhihuHttp.getZhihuHttp().getHot().subscribe(new Observer<HotJson>() {

            @Override
            public void onSubscribe(@NonNull Disposable d) {
                mDisposable = d;
            }

            @Override
            public void onNext(HotJson hotJson) {
                ArrayList<Hot> hots = hotJson.getHots();
                hotAdapter.addList(hots);
                hotAdapter.notifyDataSetChanged();
            }

            @Override
            public void onComplete() {

            }

            @Override
            public void onError(Throwable e) {
                Logger.e(e, "Subscriber onError()");
            }
        });
    }

}

File: app/src/main/java/com/yao/zhihudaily/ui/hot/HotMainFragment.kt
package com.yao.zhihudaily.ui.hot

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.orhanobut.logger.Logger
import com.yao.zhihudaily.R
import com.yao.zhihudaily.model.HotJson
import com.yao.zhihudaily.net.ZhihuApiService
import com.yao.zhihudaily.net.ZhihuHttp
import com.yao.zhihudaily.tool.SimpleDividerDecoration
import com.yao.zhihudaily.ui.BaseFragment
import io.reactivex.Observer
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.fragment_hot.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * @author Yao
 * @date 2016/7/22
 */
class HotMainFragment : BaseFragment(), SwipeRefreshLayout.OnRefreshListener {

    private var hotAdapter: HotAdapter? = null

    private var mDisposable: Disposable? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_hot, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        swipe_refresh_layout.setOnRefreshListener(this)
        rv_hots!!.layoutManager = LinearLayoutManager(fragmentActivity)
        //rv_hots.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL_LIST));
        rv_hots!!.addItemDecoration(SimpleDividerDecoration(fragmentActivity))
        hotAdapter = HotAdapter(this)
        rv_hots!!.adapter = hotAdapter
        getHot()

    }

    override fun onDestroyView() {
        super.onDestroyView()
        if (mDisposable != null) {
            mDisposable!!.dispose()
        }
    }

    override fun onRefresh() {
        Log.e("YAO", "HotMainFragment --- onRefresh()")
        getHot()
    }

    private fun getHotOriginal() {
        GlobalScope.launch(Dispatchers.Main) {

            val hotJson = withContext(Dispatchers.IO){
                val dailiesJson = ZhihuApiService.mZhihuApiService.getHotCoroutine()
                dailiesJson
            }

            if (swipe_refresh_layout.isRefreshing) {
                swipe_refresh_layout.isRefreshing = false
            }

            val hots = hotJson.hots
            if (hots != null) {
                hotAdapter!!.addList(hots)
                hotAdapter!!.notifyDataSetChanged()
            }
        }
    }

    private fun getHot() {
        GlobalScope.launch(Dispatchers.Main) {

            try {
                val hotJson = withContext(Dispatchers.IO){
                    val dailiesJson = ZhihuApiService.mZhihuApiService.getHotCoroutine()
                    dailiesJson
                }

                val hots = hotJson.hots
                if (hots != null) {
                    hotAdapter!!.addList(hots)
                    hotAdapter!!.notifyDataSetChanged()
                }
            } catch (e: Exception) {
                Log.e("YAO", e.toString())
            } finally {
                if (swipe_refresh_layout.isRefreshing) {
                    swipe_refresh_layout.isRefreshing = false
                }
            }
        }
    }

    private fun getHotBak() {
        ZhihuHttp.mZhihuHttp.getHot().subscribe(object : Observer<HotJson> {

            override fun onSubscribe(@NonNull d: Disposable) {
                mDisposable = d
            }

            override fun onNext(hotJson: HotJson) {
                val hots = hotJson.hots
                if (hots != null) {
                    hotAdapter!!.addList(hots)
                    hotAdapter!!.notifyDataSetChanged()
                }
            }

            override fun onComplete() {

            }

            override fun onError(e: Throwable) {
                Logger.e(e, "Subscriber onError()")
            }
        })
    }

    companion object {

        private val TAG = "HotMainFragment"
    }
}